email.go

```diff
@@ -188,27 +195,97 @@ func (em *Sender) buildMessage(text string, params Params) (message string, err
 	message = addHeader(message, "From", params.From)
 	message = addHeader(message, "To", strings.Join(params.To, ","))
 	message = addHeader(message, "Subject", params.Subject)
-	message = addHeader(message, "Content-Transfer-Encoding", "quoted-printable")
 
 	if em.contentType != "" {
 		message = addHeader(message, "MIME-version", "1.0")
-		message = addHeader(message, "Content-Type", fmt.Sprintf("%s; charset=%q", em.contentType, em.contentCharset))
 	}
+
 	message = addHeader(message, "Date", em.timeNow().Format(time.RFC1123Z))
 
 	buff := &bytes.Buffer{}
 	qp := quotedprintable.NewWriter(buff)
-	if _, err := qp.Write([]byte(text)); err != nil {
-		return "", err
+	mp := multipart.NewWriter(buff)
+	boundary := mp.Boundary()
+
+	withAttachments := len(params.Attachments) > 0
+	if withAttachments {
+		message = addHeader(message, "Content-Type", fmt.Sprintf("multipart/mixed; boundary=%q\r\n\r\n%s\r",
+			boundary, "--"+boundary))
 	}
-	if err := qp.Close(); err != nil {
-		return "", fmt.Errorf("failed to encode text to quoted printable: %w", err)
+
+	if em.contentType != "" {
+		message = addHeader(message, "Content-Transfer-Encoding", "quoted-printable")
+		message = addHeader(message, "Content-Type", fmt.Sprintf("%s; charset=%q", em.contentType, em.contentCharset))
+
+	}
+
+	if err := em.writeBody(qp, text); err != nil {
+		return "", fmt.Errorf("failed to write body: %w", err)
 	}
+
+	if withAttachments {
+		if err := em.writeAttachments(mp, params.Attachments); err != nil {
+			return "", fmt.Errorf("failed to write attachments: %w", err)
+		}
+	}
+
 	m := buff.String()
 	message += "\n" + m
+	// returns base part of the file location
 	return message, nil
 }
 
+func (em *Sender) readFile(path string) (fName string, fData []byte, err error) {
+	fData, err = os.ReadFile(filepath.Clean(path))
+	if err != nil {
+		return "", nil, fmt.Errorf("failed to read file %q: %w", path, err)
+	}
+	fName = filepath.Base(path)
+	return fName, fData, nil
+}
+
+func (em *Sender) writeBody(w io.Writer, text string) error {
+	qp := quotedprintable.NewWriter(w)
+	if _, err := qp.Write([]byte(text + "\r\n")); err != nil {
+		return err
+	}
+	if err := qp.Close(); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (em *Sender) writeAttachments(mp *multipart.Writer, attachments []string) error {
+	for _, attachment := range attachments {
+		fName, fData, err := em.readFile(attachment)
```

Comment:
do we really have to read the whole file to memory? Isn't encoder is a Writer? If so, we probably can do io.Copy

Reply:
If I'm not mistaken io.Copy allocates buffer im memory while reading src too. Read whole file in memory more convenient for detecting mime type
```header.Set("Content-Type", http.DetectContentType(fData)+"; name=\""+fName+"\"")```

Reply:
io.Copy does the buffering, but it won't be as big as a potentially 50M file. As far as I recall, this buffer is just 32k

regarding DetectContentType - this func doesn't need full file as it reads the first 512 bytes only

Reply:
Ok, I see.

Reply:
Done.